### PR TITLE
[language] Merge sent and received event

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

Merge the `Sent` and `Received` events into one unique event stream. The main issue to the previous two stream approach is that it's really unfriendly for the pub/sub system as the subscribe will not be able to know who the two parties of a transaction are by just looking at the event payload.

This PR will have following tradeoffs:
1. We save 36byte per account in the state store(as we only need one event handle now). In exchange the transaction execution result size will be 32byte bigger than before (as we include an extra address info in the event).
2. Two events with the same payload will be emitted during a transaction(one for sender and one for receiver). The pub/sub implementor will need to do a deduplication. 
3.  For client side, one can only get an event stream that contains all the past transactions. There would be no distinction between sent stream and received stream any more.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

cargo test

## Related PRs

#1146

### CODE REVIEW

**Comment by bmaurer** [comment]
> Two events with the same payload will be emitted during a transaction(one for sender and one for receiver). The pub/sub implementor will need to do a deduplication.

Might be worth thinking about the kinds of use cases for this payload and if this will become annoying enough that we want to put a boolean flag in to represent the send vs receive side or even  keep two event types but have both sides in the emitted event. THe two main programmer annoyances here seem to be (1) deduping if you tail all events (2) if you look at events for account X you have to reverse engineer which side of the transaction X is on.

**Comment by bmaurer** [comment]
Also: worth thinking about the kind of corner cases that might come up if an account pays itself (is this allowed?). This will result in two, indistinguishable events.

**Comment by sblackshear** [comment]
> Might be worth thinking about the kinds of use cases for this payload and if this will become annoying enough that we want to put a boolean flag in to represent the send vs receive side or even keep two event types but have both sides in the emitted event. 

The thought was that the value of your hypothetical boolean flag can be inferred by checking `tx.sender == event.payer`. We could (of course) explicitly represent the flag or separate the types (as they were before this PR) and add both parties in both events. Our conclusion was that these solutions make events/accounts bigger without making the API meaningfully easier to use. 

> THe two main programmer annoyances here seem to be (1) deduping if you tail all events

There's no de-duping here, unless I misunderstand. The event payloads will be the same, but the stream ID's and (possibly) sequence numbers will be different. Note that there are many ways to create duplicate event payloads (e.g., pay the same amount to the same recipient), so distinguishing events based on these other characteristics is important to have regardless.

> (2) if you look at events for account X you have to reverse engineer which side of the transaction X is on.

As mentioned above, this would be done by checking `tx.sender == event.payer`. It could be done with a boolean flag instead, but doesn't seem necessary. If you have a proof of an event payload, you also have a proof of its parent tx and its sender address.

**Comment by sblackshear** [comment]
> Also: worth thinking about the kind of corner cases that might come up if an account pays itself (is this allowed?). This will result in two, indistinguishable events.

This is worth thinking about more. We do currently allow self payments (and there are good reasons for doing so), but emitting two events in this case is a bit odd. We could (e.g.) choose to emit only one when the sender and recipient are the same.

Note that (similar to explanation above) the event payloads will be identical, but the events themselves will not be (different sequence number).

File: language/stdlib/modules/libra_account.mvir [hunk 1]
```diff
@@ -97,18 +83,17 @@ module LibraAccount {
...
         sender_account_ref = borrow_global_mut<T>(copy(sender));
-        // Log a send event
-        sent_event = SentPaymentEvent { amount: copy(deposit_value), payee: copy(payee) };
+        // Log a Payment event to sender
+        event = PaymentEvent { amount: copy(deposit_value), payee: copy(payee), payer: move(sender) };
 
-        Event.emit_event<Self.SentPaymentEvent>(&mut move(sender_account_ref).sent_events, move(sent_event));
+        Event.emit_event<Self.PaymentEvent>(&mut move(sender_account_ref).payment_events, copy(event));
 
         // Load the payee's account
```
Comment (sblackshear) [inline language/stdlib/modules/libra_account.mvir]: To address Ben's feedback, perhaps we could:
- Move the payee logic before the payer logic
- After executing the payee logic, check if the sender is equal to the payee and return before executing the payer logic if so.

File: language/stdlib/modules/libra_account.mvir [hunk 2]
```diff
@@ -42,39 +40,30 @@ module LibraAccount {
...
         // The address that was paid
         payee: address,
-    }
-
-    // Message for received events
-    struct ReceivedPaymentEvent {
-        // The amount of LibraCoin.T received
-        amount: u64,
-        // The address that sent the coin
+        // The address that paid libra.
```
Comment (sblackshear) [inline language/stdlib/modules/libra_account.mvir]:
```suggestion
        // The address that sent the payment
```

**Comment by runtian-zhou** [comment]
For the dedup, the problem is as following:

When the pub/sub service is tailing a transaction, it is going to see two events of identitical payload, as we are going to emit events to both sides of the parties. Thus in order to process it correctly, the pub/sub service will need to run a dedup per transaction. This wouldn't be too hard to implement as it will not be a global level dedup but only a transaction level dedup

**Comment by sblackshear** [comment]
> Thus in order to process it correctly, the pub/sub service will need to run a dedup per transaction. 

Can you elaborate on "process it correctly" means and how it goes wrong if we have duplicate payloads?

**Comment by runtian-zhou** [comment]
> > Thus in order to process it correctly, the pub/sub service will need to run a dedup per transaction.
> 
> Can you elaborate on "process it correctly" means and how it goes wrong if we have duplicate payloads?

Let's say the logic for the pub/sub service is it will notify both parties whenever it sees a Payment event. If we don't do dedup then the naive implementation will just notify both parties twice.

**Comment by sblackshear** [comment]
> Let's say the logic for the pub/sub service is it will notify both parties whenever it sees a Payment event. If we don't do dedup then the naive implementation will just notify both parties twice.

Why will it notify both parties? If I have address `0x11`, will the pub/sub system notify me anytime any event payload contains the address `0x11`?

**Comment by runtian-zhou** [comment]
> > Let's say the logic for the pub/sub service is it will notify both parties whenever it sees a Payment event. If we don't do dedup then the naive implementation will just notify both parties twice.
> 
> Why will it notify both parties? If I have address `0x11`, will the pub/sub system notify me anytime any event payload contains the address `0x11`?

That is basically the idea. A naive implementation would be exactly what you described: `0x11` will be notified whenever payload contains `0x11`. @phoenix-antigravity raised the issue to me previously and it would be nice to pull him in.

**Comment by sblackshear** [comment]
I worry that in a world with third-party publishing, this implementation strategy could be spammy. Anyone can emit an event with `0x11` in it...

I guess I would say that it makes sense to support an API that gives you all the events that contain a given address, but that this API should be use for forensics--ordinary users shouldn't subscribe to events using such an API.

**Comment by bmaurer** [comment]
> I worry that in a world with third-party publishing, this implementation strategy could be spammy. Anyone can emit an event with 0x11 in it...

But they can't emit an event of the type PaymentEvent with an address 0x11 in it.

Think of the use case here as a blockchain explorer / etc that wants to scan all transactions (vs picking an account and scanning that specific account). having 2 identical events per payment is a bit unintuitive. Not the end of the world, but it seems like it'd be a FAQ at a minimum

**Comment by phoenix-antigravity** [comment]
@sblackshear: we don't just check for address to be present in raw representation of payload. But more in a structured way. e.g. subscribe to all events of given type where one of fields(field_a or field_b) equals to target address

**Comment by runtian-zhou** [comment]
> I worry that in a world with third-party publishing, this implementation strategy could be spammy. Anyone can emit an event with `0x11` in it...
> 
> I guess I would say that it makes sense to support an API that gives you all the events that contain a given address, but that this API should be use for forensics--ordinary users shouldn't subscribe to events using such an API.

With #1261, I would imagine that the pub/sub service would just check for events with `type` field equals to `LibraAccount.PaymentEvent`.

**Comment by sblackshear** [comment]
> Think of the use case here as a blockchain explorer / etc that wants to scan all transactions (vs picking an account and scanning that specific account). having 2 identical events per payment is a bit unintuitive. Not the end of the world, but it seems like it'd be a FAQ at a minimum

Agreed, this is a definite downside of merging the event types. I'm not sure how to weigh the space savings against the unintuitive behavior in the "scan all transactions" use case. Adding the boolean flag you suggest is a good way to split the difference.

> @sblackshear: we don't just check for address to be present in raw representation of payload. But more in a structured way. e.g. subscribe to all events of given type where one of fields(field_a or field_b) equals to target address

Got it, that makes sense.

Another approach that I think could make sense is to make the `EventKey` transparent instead of a hash: expose the (creator address, counter) structure. That way, clients could subscribe to all events emitted using keys that they created. That would simplify our recommendation for clients in this case. Instead of "subscribe to PaymentEvents where recipient=me and deduplicate" for received payments or "subscribe to PaymentEvents where sender=me and tx.sender = me" for sent payments, we can simply say: "subscribe to PaymentEvents where key.creator = me". That recommendation could cover both payment sent and payment received use-cases. Furthermore, the more general "subscribe to all events where key.creator = me" could serve as a proxy for "subscribe to all events relevant to my account".

I know that some folks have expressed a preference for keeping `EventKey` opaque, but I think making it transparent is worth considering if it significantly simplifies the subscription and "deduplication" story.

**Comment by runtian-zhou** [comment]
> > Think of the use case here as a blockchain explorer / etc that wants to scan all transactions (vs picking an account and scanning that specific account). having 2 identical events per payment is a bit unintuitive. Not the end of the world, but it seems like it'd be a FAQ at a minimum
> 
> Agreed, this is a definite downside of merging the event types. I'm not sure how to weigh the space savings against the unintuitive behavior in the "scan all transactions" use case. Adding the boolean flag you suggest is a good way to split the difference.
> 
> > @sblackshear: we don't just check for address to be present in raw representation of payload. But more in a structured way. e.g. subscribe to all events of given type where one of fields(field_a or field_b) equals to target address
> 
> Got it, that makes sense.
> 
> Another approach that I think could make sense is to make the `EventKey` transparent instead of a hash: expose the (creator address, counter) structure. That way, clients could subscribe to all events emitted using keys that they created. That would simplify our recommendation for clients in this case. Instead of "subscribe to PaymentEvents where recipient=me and deduplicate" for received payments or "subscribe to PaymentEvents where sender=me and tx.sender = me" for sent payments, we can simply say: "subscribe to PaymentEvents where key.creator = me". That recommendation could cover both payment sent and payment received use-cases. Furthermore, the more general "subscribe to all events where key.creator = me" could serve as a proxy for "subscribe to all events relevant to my account".
> 
> I know that some folks have expressed a preference for keeping `EventKey` opaque, but I think making it transparent is worth considering if it significantly simplifies the subscription and "deduplication" story.

My concern is making the `EventKey` transparent will not solve the problem for now. The problem is for account creation, we are still using the sender's address to create the `EventKey` rather than the fresh address. And even if we solve the problem, this solution might only work with the `Payment` event that we have.

**Comment by sblackshear** [comment]
> My concern is making the EventKey transparent will not solve the problem for now. The problem is for account creation, we are still using the sender's address to create the EventKey rather than the fresh address.

Ah, I was assuming that this would be solved once https://github.com/libra/libra/pull/1208 lands. Is this not the case?

> And even if we solve the problem, this solution might only work with the Payment event that we have.

I contend that this is (or at least, can be) a very general-purpose solution. We should encourage a programming discipline where resources that belong to me use `EventKey`'s created by me to emit events. This is already the natural way to do things given the restrictions on creating event handles. For example:

```
resource T { event_handle: Event.Handle<Something>, ... }

public create1() {
    move_to_sender<T>(T { event_handle: EventHandle.create<Something>, ... })
}

public create2(event_handle: Event.Handle<Something>) {
  assert(event_handle.creator() == tx.sender);
  move_to_sender<T>(T { event_handle, ... })
}
```

By contrast, heuristics like filtering events based on the contents of a specific field in a specific event type (e.g.  "subscribe to PaymentEvents where sender=me and tx.sender = me") are (by definition) not a general-purpose solution. I'm sure we'll need some heuristics like this in some cases. But creating a separate subscription heuristic for each new event type doesn't seem tenable to me.

**Pull request closed without merging**